go/chat/server.go

```diff
@@ -3822,3 +3827,200 @@ func (h *Server) ToggleEmojiAnimations(ctx context.Context, enabled bool) (err e
+				IdentifyBehavior:  arg.IdentifyBehavior,
+				EphemeralLifetime: ephemeralLifetime,
+			},
+		})
+	default:
+		return h.PostLocal(ctx, chat1.PostLocalArg{
+			SessionID:      arg.SessionID,
+			ConversationID: arg.DstConvID,
+			Msg: chat1.MessagePlaintext{
+				ClientHeader: chat1.MessageClientHeader{
+					Conv:              dstConv.Info.Triple,
+					TlfName:           dstConv.Info.TlfName,
+					TlfPublic:         dstConv.Info.Visibility == keybase1.TLFVisibility_PUBLIC,
+					MessageType:       mvalid.ClientHeader.MessageType,
+					EphemeralMetadata: mvalid.EphemeralMetadata(),
+				},
+				MessageBody: mvalid.MessageBody.DeepCopy(),
+			},
+			IdentifyBehavior:   arg.IdentifyBehavior,
+			SkipInChatPayments: true,
+		})
+	}
+}
+
+func (h *Server) ForwardMessageNonblock(ctx context.Context, arg chat1.ForwardMessageNonblockArg) (res chat1.PostLocalNonblockRes, err error) {
+	var identBreaks []keybase1.TLFIdentifyFailure
+	ctx = globals.ChatCtx(ctx, h.G(), arg.IdentifyBehavior, &identBreaks, h.identNotifier)
+	defer h.Trace(ctx, &err, "ForwardMessageNonblock")()
+	defer h.suspendBgConvLoads(ctx)()
+	defer func() { h.setResultRateLimit(ctx, &res) }()
+	uid, err := utils.AssertLoggedInUID(ctx, h.G())
+	if err != nil {
+		return res, err
+	}
+	reason := chat1.GetThreadReason_FORWARDMSG
+	msg, err := h.G().ConvSource.GetMessage(ctx, arg.SrcConvID, uid, arg.MsgID, &reason, nil, true)
+	if err != nil {
+		return res, err
+	} else if !msg.IsValid() {
+		return res, fmt.Errorf("unable to forward message, source is invalid")
+	}
+	dstConv, err := utils.GetVerifiedConv(ctx, h.G(), uid, arg.DstConvID, types.InboxSourceDataSourceAll)
+	if err != nil {
+		return res, err
+	}
+	mvalid := msg.Valid()
+	switch mvalid.ClientHeader.MessageType {
+	case chat1.MessageType_ATTACHMENT:
+		mbod := msg.Valid().MessageBody.Attachment()
+		dl, err := h.DownloadFileAttachmentLocal(ctx, chat1.DownloadFileAttachmentLocalArg{
+			SessionID:        arg.SessionID,
+			ConversationID:   arg.SrcConvID,
+			MessageID:        arg.MsgID,
+			DownloadToCache:  true,
+			IdentifyBehavior: arg.IdentifyBehavior,
+		})
+		if err != nil {
+			return res, err
+		}
+		var ephemeralLifetime *gregor1.DurationSec
+		if md := mvalid.EphemeralMetadata(); md != nil {
+			ephemeralLifetime = &md.Lifetime
+		}
+		res, err := h.PostFileAttachmentLocalNonblock(ctx, chat1.PostFileAttachmentLocalNonblockArg{
+			SessionID: arg.SessionID,
+			Arg: chat1.PostFileAttachmentArg{
+				ConversationID:    arg.DstConvID,
+				TlfName:           dstConv.Info.TlfName,
+				Visibility:        dstConv.Info.Visibility,
+				Filename:          dl.FilePath,
+				Title:             mbod.Object.Title,
+				Metadata:          mbod.Metadata,
+				IdentifyBehavior:  arg.IdentifyBehavior,
+				EphemeralLifetime: ephemeralLifetime,
+			},
+		})
+		if err != nil {
+			return res, err
+		}
+		go func() {
+			defer os.Remove(dl.FilePath)
+			outbox := storage.NewOutbox(h.G(), uid)
```

Comment:
Use GetUploadTempFile

Reply:
@mmaxim not sure how that applies here, what do you mean exactly? we download the file first and then just cleanup in this routine.